python/ray/data/_internal/compute.py

```diff
@@ -254,15 +254,30 @@ def _apply(
 
         if name is None:
             name = "map"
-        blocks_in = block_list.get_blocks_with_metadata()
-        # Bin blocks by target block size.
+        blocks_in: List[
+            Tuple[ObjectRef[Block], BlockMetadata]
+        ] = block_list.get_blocks_with_metadata()
+
+        # We bundle blocks according to the following rules:
+        # 1. Attempt to bundle up to the target block size.
+        # 2. If the max concurrency of the ActorPool is set, then
+        #    cap the number of bundles to match the size of the ActorPool.
+        #    This avoids additional overhead in submitting new actor tasks and allows
+        #    the actor task to do optimizations such as batch prefetching.
+        target_num_bundles = float("inf")
         if target_block_size is not None:
             _check_batch_size(blocks_in, target_block_size, name)
-            block_bundles = _bundle_blocks_up_to_size(
-                blocks_in, target_block_size, name
-            )
+            total_size = sum(metadata.num_rows for _, metadata in blocks_in)
+            target_num_bundles = min(target_num_bundles, total_size / target_block_size)
+        target_num_bundles = min(target_num_bundles, self.max_size)
+        if not math.isinf(target_num_bundles):
+            target_block_size = total_size // target_num_bundles
+            block_bundles: List[
+                Tuple[Tuple[ObjectRef[Block]], Tuple[BlockMetadata]]
+            ] = _bundle_blocks_up_to_size(blocks_in, target_block_size)
```

Comment:
Should be able to do the min/capping on the target block size itself, which should simplify this logic a bit.
```python
        # We bundle blocks according to the following rules:
        # 1. Attempt to bundle up to the target block size.
        # 2. If the max concurrency of the ActorPool is set, then
        #    cap the number of bundles to match the size of the ActorPool.
        #    This avoids additional overhead in submitting new actor tasks and allows
        #    the actor task to do optimizations such as batch prefetching.
        if target_block_size is not None:
            _check_batch_size(blocks_in, target_block_size, name)
            total_size = sum(metadata.num_rows for _, metadata in blocks_in)
            target_block_size = min(target_block_size, total_size // self.max_size)
            block_bundles: List[
                Tuple[Tuple[ObjectRef[Block]], Tuple[BlockMetadata]]
            ] = _bundle_blocks_up_to_size(blocks_in, target_block_size)
```

Reply:
thats right good point!

Reply:
Cleaned up a bit more, but don't think it can be fully simplified that much since we want to still use `self.max_size` even if `target_block_size` is None.

Reply:
Ah that's right! In that case, we should also move the logged `_check_batch_size()` to after we've modified the `target_block_size` according to the max actor pool size (and `_check_batch_size()` should probably be renamed to `_check_target_block_size()`, since that's more accurate).

How about this, which is pretty similar to what you have?
```python
        # We bundle blocks according to the following rules:
        # 1. Attempt to bundle up to the target block size.
        # 2. If the max concurrency of the ActorPool is set, then
        #    cap the number of bundles to match the size of the ActorPool.
        #    This avoids additional overhead in submitting new actor tasks and allows
        #    the actor task to do optimizations such as batch prefetching.
        if target_block_size is None:
            target_block_size = 0
        if not math.isinf(self.max_size):
            total_size = sum(
                meta.num_rows if meta.num_rows is not None else 0
                for _, meta in blocks_in)
            pool_max_block_size = total_size // self.max_size
            target_block_size = max(target_block_size, pool_max_block_size)
        if target_block_size > 0:
            _check_target_block_size(blocks_in, target_block_size, name)
            block_bundles: List[
                Tuple[Tuple[ObjectRef[Block]], Tuple[BlockMetadata]]
            ] = _bundle_blocks_up_to_size(blocks_in, target_block_size)
```
This is basically what you have except:
1. Have `target_block_size = 0` signal no bundling, since bundling blocks to size 0 doesn't make sense and would always produce empty bundles, and map `target_block_size is None` to this.
2. If the number of rows for a block is not available, fallback to 0. Falling back to more conservative bundling when encountering an unknown sized block instead bundling all blocks into a single bundle seems better to me.
3. Check the target block size after applying the actor pool size bound.

Reply:
Good simplification! Updated the code to this, thanks!